tools/wptserve/wptserve/request.py

```diff
@@ -613,14 +632,18 @@ def __init__(self, headers):
 
         if "authorization" in headers:
             header = headers.get("authorization")
-            assert isinstance(header, text_type)
-            auth_type, data = header.split(" ", 1)
+            assert isinstance(header, binary_type)
+            auth_type, data = header.split(b" ", 1)
+            auth_type = maybedecode(auth_type)
             if auth_type in auth_schemes:
                 self.username, self.password = auth_schemes[auth_type](data)
             else:
                 raise HTTPException(400, "Unsupported authentication scheme %s" % auth_type)
 
     def decode_basic(self, data):
-        assert isinstance(data, text_type)
-        decoded_data = base64.decodestring(data.encode("utf-8"))
-        return decoded_data.decode("utf-8").split(":", 1)
+        assert isinstance(data, binary_type)
+        # RFC7617 Section 2.1 does not specify the encoding for username &
+        # passsword (as long it's compatible with ASCII). UTF-8 should be a
+        # relatively safe choice as Firefox and other browsers use it.
+        decoded_data = base64.b64decode(data).decode("utf-8")
```

Comment:
Given the historic differences here, I think we should give the username/password as bytes? Certainly we shouldn't do something irreversible (and throw in the invalid UTF-8 case).